---
title: "Preliminary RNAseq analysis"
author: "Jieun Jeong"
date: "`r Sys.Date()`"
output: pdf_document
---
This example analyzes RNAseq data from GEO data series GSE263611.  In this 
project, four cell lines of cardiomyocytes derived from patients' blood
are cultured in two ways: control and treated with IFNg.  Thus we observe
expression variability stemming from initial genetic and epigenetic differences
and from INFg treatment.

Input tables for processing (count data, meta data) are created with separate
script, shown in another document.  The original read count data uses Ensembl 
gene identifiers used in mapping fragments to genes in this project.  We
convert identifiers as needed later and reformat as a matrix with row names.

```{r warning=F, messages=F}
library(tinytex)
library(tidyverse)
pa <- function(x) file.path('/Users/jieun/Work/Git_test/RNAseq_limma',x)
# making input tables
hugo <- readRDS(pa('HUGO.RDS'))
samples <- readRDS(pa('GSE263611_key.RDS')) %>% filter(Assay == "RNA-seq")
print(samples)
types_of_interest <- c("gene with protein product", "pseudogene", "RNA,
                       long non-coding")
counts <- readRDS(pa('GSE263611_counts.RDS')) %>% left_join(hugo, by="Ensembl") %>%
  filter(Type %in% types_of_interest) %>% select(-Type)

countsAS <- counts %>% filter(Symbol != "") %>% distinct(Symbol, .keep_all = T) %>%
  select(-Entrez, -Ensembl)
countsAS <- as.matrix(column_to_rownames(countsAS, "Symbol"))

countsN <- counts %>% filter(Entrez != "") %>% distinct(Entrez, .keep_all = T) %>%
  select(-Symbol, -Ensembl)
countsN <- as.matrix(column_to_rownames(countsN, "Entrez"))
```
As the next step, we normalize count data and check the distribution of values.
```{r warning=F, messages=F}
library(edgeR)
### Definitions for limma
treatment <- as.factor(samples$Treatment)
cell_line <- as.factor(samples$CellLine)
design <- model.matrix(~ 0 + treatment) # another design will be used too
contrast <- makeContrasts(treatmentIFNg - treatmentControl, levels = design)

# Normalization for limma
dge <- DGEList(counts = countsAS) # TMM normalization, reformmating
dge <- calcNormFactors(dge)
v <- voom(countsAS, design, plot = T) # log normalization and adjustment
# checking the normalized distribution of gene expression
A <- as.data.frame(v$E)
A <- rownames_to_column(A,var='gene')
AL <- pivot_longer(A,-gene,names_to = "sample", values_to = "logExpr")
AL$sample <- as.factor(AL$sample)
p <- ggplot(AL,aes(x=sample,y=logExpr)) + geom_violin(trim=F,width=0.8)
p + stat_summary(fun.data=mean_sdl, geom="pointrange", color="red",fun.args=list(mult=1))
```
Other useful visualization give highlights of the effect of the treatment,
i.e. INFg.  First we compute a new table
```{r warning=F, messages=F}
### Clustering on v$E
library(SummarizedExperiment)
sample_info <- samples %>% select(-Assay, -Accession)
se <- SummarizedExperiment(assays = list(counts = v$E),
                           colData = sample_info)
colData = as.data.frame(colData(se))

```
then we see the heatmap
```{r warning=F, messages=F}
library(pheatmap)
# Select the top 100 most variable genes among the samples
VG <- apply(v$E, 1, var)
selectedGenes <- names(VG[order(VG, decreasing = T)])[1:100]

p <- pheatmap(v$E[selectedGenes, ],
         scale = 'row',
         show_rownames = FALSE,
         annotation_col = colData)
```
and finally, PCA results
```{r warning=F, messages=F}
library(factoextra) # for PCA

df_t <- t(v$E)

# Using only most variable genes
pr1 <- prcomp(df_t[,selectedGenes], scale. = TRUE)

p <- fviz_pca_ind(pr1,
             geom.ind = "point",
             col.ind = colData$Treatment,
             palette = "jco",
             addEllipses = TRUE,
             legene.title = "Treatment")
p
```
The heatmap shows that some genes have high fold change explained by Treatment
and some my the difference between cell lines UDID006 and UDID088
(S26, S27, S30, S31) and the other two.  PCA plot groups by Treatment, but we
can also check how cell line look in two principal PCA components
```{r warning=F, messages=F}
pca_coords <- as.data.frame(pr1$x)
pca_coords$Sample <- rownames(pca_coords)
pca_coords$CellLine <- colData$CellLine
pca_coords$Treatment <- colData$Treatment  # Add treatment groups
pca_mid <- pca_coords %>%
  group_by(CellLine) %>%
  summarise(
    x_mid = mean(PC1),
    y_mid = mean(PC2)
  )
p <- ggplot(pca_coords, aes(x = PC1, y = PC2, color = Treatment)) +
  geom_point(size = 4) +
  # stat_ellipse(type = "norm", level = 0.95) +  # 95% confidence ellipses
  geom_line(aes(group = CellLine), color = "gray") +
  geom_text(data = pca_mid, aes(x=x_mid, y=y_mid, label=CellLine), color="black", size=3) +
  
  xlim(-10, 10) + ylim(-10, 10) +
  theme_minimal()
p
```
